specs/networking/libp2p-standardization.md

```diff
@@ -0,0 +1,192 @@
+propagation times and message duplication. Current network-related defaults are:
+
+```
+(
+	// The target number of peers in the overlay mesh network (D in the libp2p specs).
+	mesh_size: 6
+	// The minimum number of peers in the mesh network before adding more (D_lo in the libp2p specs).
+	mesh_lo: 4
+	// The maximum number of peers in the mesh network before removing some (D_high in the libp2p sepcs).
+	mesh_high: 12
+	// The number of peers to gossip to during a heartbeat (D_lazy in the libp2p sepcs).
+	gossip_lazy: 6 // defaults to `mesh_size`
+	// Time to live for fanout peers (seconds).
+	fanout_ttl: 60
+	// The number of heartbeats to gossip about.
+	gossip_history: 3
+	// Time between each heartbeat (seconds).
+	heartbeat_interval: 1
+)
+```
+
+### Topics
+
+*The Go and Js implementations use string topics - This is likely to be
+updated to topic hashes in later versions - https://github.com/libp2p/rust-libp2p/issues/473*
+
+For Eth2.0 clients, topics will be sent as `SHA2-256` hashes of the topic string.
+
+There are two main topics used to propagate attestations and beacon blocks to
+all nodes on the network.
+
+- The `beacon_block` topic - This topic is used solely for propagating new
+	beacon blocks to all nodes on the networks.
+- The `beacon_attestation` topic - This topic is used to propagate
+	aggregated attestations to subscribing nodes (typically block proposers) to
+	be included into future blocks. Attestations will be aggregated in their
+	respective subnets before publishing on this topic.
+
+Shards will be grouped into their own subnets (defined by a shard topic). The
+number of shard subnets will be defined via `SHARD_SUBNET_COUNT` and the shard
+`shard_number % SHARD_SUBNET_COUNT` will be assigned to the topic:
+`shard{shard_number % SHARD_SUBNET_COUNT}`.
+
+### Messages
+
+#### Libp2p Specification
+
+*This section simply outlines the data sent across the wire as specified by
+libp2p - this section is aimed at gossipsub implementers to standardize their implementation of this protocol*
+
+Libp2p raw gossipsub messages are sent across the wire as fixed-size length-prefixed byte arrays.
+
+The byte array is prefixed with an unsigned 64 bit length number encoded as an
+`unsigned varint` (https://github.com/multiformats/unsigned-varint). Gossipsub messages therefore take the form:
+```
++--------------------------+
+|     message length       |
++--------------------------+
+|                          |
+|       body (<1M)         |
+|                          |
++--------------------------+
+```
+
+The body represents a protobuf-encoded [Message](https://github.com/libp2p/go-libp2p-pubsub/blob/master/pb/rpc.proto#L17-L24). 
+
+In the following section we discuss the data being sent in the `data` field of
+the protobuf gossipsub `Message`.
+
+#### Eth2.0 Specifics
+
+Each message has a maximum size of 512KB (estimated from expected largest uncompressed
```

Comment:
It's worth specifying that this is Eth2.0's choice, as gossipsub does not impose a max message length, to my knowledge.

Reply:
Speaking of the maximum size of a message, even though there isn't a specified constant, I played with the max size of [`DelimitedReader`](https://github.com/libp2p/go-libp2p-pubsub/blob/master/comm.go#L31), and now suspect it is the upper bound of the size in go implementation, IIUC.

Reply:
Floodsub in rust-libp2p has a maximum size of 2048 bytes:
https://github.com/libp2p/rust-libp2p/blob/master/protocols/floodsub/src/protocol.rs#L60
I've made this configurable in the gossipsub implementation. 
The max size exists so that if we see a large amount of data coming through a stream we only read a maximum amount. I understand this to prevent potential DOS vectors from malicious user's sending arbitrary large streams of data.